CodeModel/CodeParsers/CSharp/ClassCohesionAnalyzer/CohesionAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using CodeModel.CaDETModel.CodeItems;
+
+namespace CodeModel.CodeParsers.CSharp.ClassCohesionAnalyzer
+{
+    public class CohesionAnalyzer
+    {
+        public List<CohesiveParts> IdentifyCohesiveParts(CaDETClass parsedClass)
+        {
+            ClassInteractions classInteractions = new ClassInteractions(parsedClass);
+
+            IEnumerable<CohesiveParts> possibleParts = GetAllPossibleParts(classInteractions);
+
+            return possibleParts.Where(cohesiveParts =>
+                    cohesiveParts.Parts.TrueForAll(set => CalculateLCOM(set) < 0.4))
+                .ToList();
+        }
+
+        private IEnumerable<CohesiveParts> GetAllPossibleParts(ClassInteractions classInteractions)
+        {
+            var accessesThatCannotBeCut = classInteractions.GetAccessesThatCannotBeCut();
+            var accessesToBeCut = GetAccessesThatCanBeCut(classInteractions.Accesses, accessesThatCannotBeCut);
+            var possibleParts = new List<CohesiveParts>();
+            foreach (var accesses in accessesToBeCut)
+            {
+                if (possibleParts.Any(s => accesses.IsProperSupersetOf(s.AccessesToCut))) continue;
+                var parts = FindValidCohesiveParts(accesses, new HashSet<Access>(classInteractions.Accesses));
+                if (parts is not { Count: 2 }) continue;
+                var cohesiveParts = new CohesiveParts(accesses, parts);
+                possibleParts.RemoveAll(s => s.AccessesToCut.IsProperSupersetOf(cohesiveParts.AccessesToCut));
+                possibleParts.Add(cohesiveParts);
+            }
+
+            return possibleParts;
+        }
+
+        private IEnumerable<HashSet<Access>> GetAccessesThatCanBeCut(HashSet<Access> allAccesses,
+            IEnumerable<HashSet<Access>> accessesThatCannotBeCut)
+        {
+            var data = allAccesses.ToArray();
+            if (data.Length == 0) return new List<HashSet<Access>>();
+
+            var allAccessesCombinations = Enumerable
```

Comment:
This is a heavy expression that is difficult to understand.

Reply:
I have created my own implementation of this method that is slower than this one. That's the reason I left it out. Implementation was a loop iterating through a number of elements that we want in a group (0, n/2 - 1). Inside the loop is a call to a recursive function that finds all combinations for a given number of elements. It should be faster with caching but I am not sure if it'll be as fast as this.
Any suggestions?
Should I put this in a function and write some inline comments explaining some details?

Reply:
A comment might be sufficient. It's ok to have complex code when there is a significant performance gain, and performance is an issue with this algorithm.